### ISSUE TITLE
Chekout should not complete order on payment failed

### ISSUE THREAD

**Issue by antonioperic** [Sylius/Sylius#10729]
At the moment we have a situation if you are using a third-party payment gateway, and you redirect the user to the payment provider you are losing your cart. So if the user clicks the back button, or cancel they will be redirected to the sylius shop but they cart will be lost.

Is there any suggested solution to keep the cart on the back button, and is there any plan to fix this in future? 

From my POV I think we should have a simple ability to complete the order only on successful payment.

**Comment by stefandoorn** [comment]
The issue is indeed present - I heard clients also complaining about it. Another issue is here probably that not everyone visits the confirmation page and sometimes 'paid' notifications from payment providers are slightly delayed - but there is probably a good solution to this.

**Comment by gabiudrescu** [comment]
we have the same issue on our website, but because it happens quite rarely, we decided to give it low prio. 

but as we grow bigger, it becomes more common and at a certain point, it should be addressed. 

the problem comes from the way Sylius is handling payments, especially with payment service providers that require redirect of customers on their 3rd party payment pages. 

when a customer clicks "Send order" in Sylius, the order is transitioned from cart to order and then sent to 3rd party payment service provider page. if the customer hits back, it is normal, following current logic and implementation to see an empty cart.

ideally, the order should be transitioned from cart to order only after the customer came back from the payment service provider - in the thank you page. but this could imply a great deal of BC Breaks in Sylius. though, if you put yourself in the shoes of the customer, it might not be that bad after all. 

another solution I might be thinking without having to change the entire order / cart state machine is to implement the following logic:
 - cart
 - complete order
 - redirect to an intermediary page where you show the customers the order that has been completed and they should wait a second to be redirected for payment
 - redirect them to payment service provider
 - if they hit back from the browser, detect this in javascript, on browser level and keep them in the page where you show them the completed order and you allow them to either retry payment or change payment method

of course, this does not help for the usecase when someone hits back because they want to remove a product from cart or if they want to change qtys. if you want to solve this, you need to implement the previous solution.

**Comment by dbalabka** [comment]
The major issue, in this case, is that item will be reserved and not will be available for order. It means that even if customer tries to add item into cart once again, it might be not possible.
In any scenario, there is a very high probability to lose the customer.

**Comment by dbalabka** [comment]
I see the following possible solutions:
1. Reset the state of Order. It is not so easy because you have to deal with Payment, Shipment state as well. It might lead to some issues.
2. Much safer solution is to use [Reorder Plugin](https://github.com/Sylius/CustomerReorderPlugin) with after order fails or has been cancelled. It would be BC safe change. Also, the history of failing/cancelled order will persist. It is complicated to catch a case when a customer just presses the back button. I guess we can keep current Order in session and restore when it is impossible to fetch cart.
3. Another and probably, the most bulletproof solution is to tweak StateMachine. So the order should have only `cart` and `success` states.

**Comment by dbalabka** [comment]
@CoderMaggie is it possible to increase the priority of this ticket?

**Comment by dbalabka** [comment]
So I have implemented the second option from [proposed](https://github.com/Sylius/Sylius/issues/10729#issuecomment-546685929). Currently, it works but it is very hacky. As I see the only way to solve it properly is to review the state machine states and keep order in `cart` state until it will be paid or all items from order will be removed (the third option form [comment](https://github.com/Sylius/Sylius/issues/10729#issuecomment-546685929) above)

**Comment by antonioperic** [comment]
just raising this cc @coderMaggie

**Comment by antonioperic** [comment]
not stale

**Comment by gabiudrescu** [comment]
ping @CoderMaggie

**Comment by vvasiloi** [comment]
What about changing the `Sylius\Bundle\CoreBundle\Checkout\CheckoutResolver` and instead of just redirecting to cart page
```
        if ($order->isEmpty()) {
            $event->setResponse(new RedirectResponse($this->urlGenerator->generateForCart()));
        }
```
check if session has `sylius_order_id`, then get the order and redirect to `sylius_shop_order_show`

**Comment by dbalabka** [comment]
@vvasiloi now sure that it will solve the problem with a lost cart. Customer will be redirected to a page with only order information. As I described in the above comments, the customer should be able to proceed with checkout once again.

Anyway redirect will not cover other customers behaviour like just pressing back button from the third-party payment page, that lead to an empty cart also.

**Comment by vvasiloi** [comment]
> pressing back button

@dbalabka that's exactly for this scenario, they will be redirected to the page where the payment page.
Adding a check like that to the cart page will fix the "lost cart" problem as well.

**Comment by dbalabka** [comment]
@vvasiloi am I right that customer:
1. **will be able** only to change the payment method 
2. and **not be able** to resume the checkout process 
3. and **not be able** to change the order items 

in your scenario?

Also, interesting what will happen if the customer will try to directly open the cart page in separates browser's tab during a payment process in another separate tab (it is a real-life scenario).

**Comment by vvasiloi** [comment]
@dbalabka yes, all correct. After the order was placed, it can't be modified by the customer.
Accessing any checkout page, but not cart page, after order is placed will redirect to this page:
![image](https://user-images.githubusercontent.com/7114562/82545196-fea9e680-9b5e-11ea-9236-e45f76fca0e3.png)

**Comment by dbalabka** [comment]
@vvasiloi IMO it is a simple and quick workaround but unfortunately, do not the solve the problem. As I described above, the more correct solution (and most complicated due to BC break) is to rid of order’s `new` state. 

It is important to give the possibility for the customer to be able to navigate between checkout steps back and forward (to complete steps) or jump between steps in any order. For example, such an approach provides the possibility to handle cart abandonment that might happen on any checkout step even after the failed payment. Customer should be able to return to the cart to make some changes (e.g. apply coupon) and proceed with checkout once again.

**Comment by jordisala1991** [comment]
Is there any workaround for this issue?

**Comment by dbalabka** [comment]
@jordisala1991 I have tried second from this list: https://github.com/Sylius/Sylius/issues/10729#issuecomment-546685929
I’m not suggesting to waist you time on it because it is complicated and requires a lot of development effort.

IMO we should wait when the third one from the list above will be implemented. `new` state is really redundant and introduce this issue.

**Comment by vvasiloi** [comment]
@dbalabka I don't think the `new` state is the issue here.
What is relevant is not the state itself, but the transition.
The order is `created` (`sylius_order` transition from `cart` to `new`) when the checkout is completed (`sylius_order_checkout` transition from `payment_selected`/`payment_skipped` to `complete`).
By default this happens when the customer submits the form to `sylius_shop_checkout_complete` ([route config](https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/ShopBundle/Resources/config/routing/checkout.yml#L81-L83)).
If someone does not want the checkout to be completed if the payment is not successful, then the state machine transition should be removed from this route and done somewhere else, like in a `sylius_payment` state machine transition callback.
Of course, I oversimplified and more changes are required, especially to payment and order state machine callbacks.
A detailed diagram that also includes cascade transitions will certainly help better understand the implications, but I can't do one right now.
An issue I have right now is that quite often customers don't wait for the redirect from payment providers and in a scenario like that the checkout won't be completed and the order won't be created.
Ideally, I would like the checkout flow to easily allow inclusion the external payment (or other) step(s), instead of completing right before these, but also easily allow the current flow for scenarios like paying with cash on delivery, invoice or not having to pay at all.

**Comment by dbalabka** [comment]
@vvasiloi I appreciate your input to solve this problem. Still, I think that order’s State Machine might have less steps. I would like find solution that will be bulletproof and cover more edge cases as possible.
One of them is that you have mentioned:
> An issue I have right now is that quite often customers don't wait for the redirect from payment providers and in a scenario like that the checkout won't be completed and the order won't be created.

Could you please share your thoughts about possible reliable solutions?

> Ideally, I would like the checkout flow to easily allow inclusion of external payment (or other) steps, instead of completing right before these, but also easily allow the current flow for scenarios like paying with cash on delivery, invoice or not having to pay at all.

Fully agree. Currently, it is complicated to integrate PayPal with Payments API to authorize a payment that you can capture later.

**Comment by kiakahaDZ** [comment]
Hi everyone i want to keep the content of cart after payment is failed in my case it delete all the cart and his items if the payment is failed how i can do that plz?

**Comment by David-Crty** [comment]
Does someone manage to handle this issue on a live project?
I'm surprised that very few peoples have this issue has it look to be a common e-commerce issue?

**Comment by vvasiloi** [comment]
@David-Crty the order is not completed if the payment fails, only the checkout is completed. When the payment fails, businesses usually allow the customer to retry the payment in a more proactive way than Sylius currently does.

**Comment by ajdinmore** [comment]
How is this still open?  ~~As is, the software isn't fit for purpose for the vast majority, if not all, of our clients.~~

ETA: Quite embarrassed myself here; I've just found the reason I flew off the handle was because I'd told our designer they could replace the order complete page, without checking what was there first.  Turns out that's where failed/cancelled payments are handled, so there's a way smaller problem here than I thought. 😳

**Comment by vvasiloi** [comment]
@ajdinmore would you like to contribute or suggest a solution or anything meaningful that would help close this issue?

**Comment by ajdinmore** [comment]
Once I've finished looking at it, maybe.  But for something this major, if a backwards compatible workaround can't be worked out in nearly four years, I'd say it's time for a new major version branch so you can have some BC breaks.

Sorry though; I should apologise for my outburst, it was totally uncalled for.

**Comment by vvasiloi** [comment]
It's not the backward compatibility or the lack of workarounds that are preventing us from solving this issue.
Every project I know solved this issue in a different way to better fit their business processes.
So the problem is to come up with a solution that's good enough for most projects and is also at least as open (not restrictive) as the current one.
You say that it doesn't fit your clients, so what does fit?
At least describe the desired behaviour from the business perspective.

**Comment by leflings** [comment]
In my team we were able to come up with a workaround for this problem by adding a new state to the state machine. But it involves completely overriding the SM configurations, because we have to remove some transitions. There are furthermore some changes needed to PayumController and some other Payum related classes (in Sylius scope - not Payum-Payum scope).

I've been meaning to do a write up of this - both for internal documentation and to share here for others encountering the same issue.

The solution is not very satisfying and it involves manually inspecting related files whenever you upgrade Sylius (minor) versions, but it does work.

I think by sharing our solution, maybe it can be identified which Sylius implementa needs to be extendable for a solution to be possible in a less icky manner.

Currently very busy at new job, so uncertain when/if I have time to sit down and do a write up.

**Comment by vvasiloi** [comment]
@leflings Thanks, I hope you'll find time to share it!

**Comment by ajdinmore** [comment]
@vvasiloi I think the easiest way to describe the desired behaviour is that the customer should be able to go back and alter their order at any point up until a payment has been confirmed, whether the payment process has been initiated or not.

@leflings You mean the payment state machine?  We're already overriding the checkout state machine, so that process isn't a concern.  

From what I've seen so far, I'm guessing the easiest option is going to be adding a failed/cancelled state which then repopulates the cart from the incomplete order and trashes that order.

**Comment by vvasiloi** [comment]
@ajdinmore as a counter-example, this is the opposite of how it should work for payment methods like cash-on-delivery, bank transfer, financing and any other forms of payment that require placing an order first.

**Comment by ajdinmore** [comment]
Yes, quite true.  I'm just surprised the situation is the way it is as I can't recall a single company we've worked with that hasn't primarily used payment at the time of order.  Maybe it's a regional thing.

**Comment by vvasiloi** [comment]
All the projects I worked on so far are EU-based (none in UK though) and all of them allow the customer to place orders even if the online payment fails and they would then allow the customer to retry the payment later with the same or a different payment method.
The general idea is to make it as easy as possible for the customer to place an order and all the issues are solved afterwards one way on another.

**Comment by vvasiloi** [comment]
Of course, the ideal solution should allow both scenarios and maybe more. It should perhaps be configurable at the payment method level or channel or a combination of both.

**Comment by ajdinmore** [comment]
If the payment fails, a retry with the same order would be expected, yes, but immediately.  From the people I've spoken to about this the expectation seems to be that, from the customer's perspective, there would be no order and the items would not have left the cart when this occurs; they would be exactly as they were at the previous step of the checkout process - payment selection.

I'm not going to discuss it too much before I've had a chance to dig into it properly, but my gut says the payment method level would be sufficient; if necessary, the method could be duplicated to provide the option at the channel level.

**Comment by lusimeon** [comment]
On my side I've added a transition `reset` to state machines, these transitions reset state to `cart` for the order related states machines (sylius_order_checkout, sylius_order, sylius_order_payment, sylius_order_shipping and sylius_shipment).
Transitions are then triggered if users leave the checkout after `completed` state (based on a whitelist of Symfony routes) and payment method is not offline. This part should be improved to be more flexible.
Then you have to add some logic to revert behaviors introduce by checkout `complete` transition (e.g. `sylius_order` `reset` transition should release inventory).

Maybe this not fit for all payment method (e.g. split payment).
